File: esap-core/src/main/java/ru/javavlsu/kb/esap/service/UserDeviceTokenService.java
package ru.javavlsu.kb.esap.service;

import org.springframework.stereotype.Service;
import ru.javavlsu.kb.esap.dto.notifications.TokenRequest;
import ru.javavlsu.kb.esap.exception.NotFoundException;
import ru.javavlsu.kb.esap.model.TokenStatus;
import ru.javavlsu.kb.esap.model.User;
import ru.javavlsu.kb.esap.model.UserDeviceToken;
import ru.javavlsu.kb.esap.repository.UserDeviceTokenRepository;

@Service
public class UserDeviceTokenService {
    private final UserDeviceTokenRepository userDeviceTokenRepository;

    public UserDeviceTokenService(UserDeviceTokenRepository userDeviceTokenRepository) {
        this.userDeviceTokenRepository = userDeviceTokenRepository;
    }

    public void saveToken(User user, TokenRequest request) {
        String tokenValue = request.getToken();

        if (!userDeviceTokenRepository.existsByTokenAndUser(tokenValue, user)) {
            UserDeviceToken deviceToken = new UserDeviceToken();
            deviceToken.setToken(tokenValue);
            deviceToken.setUser(user);
            deviceToken.setStatus(TokenStatus.ACTIVE);
            userDeviceTokenRepository.save(deviceToken);
        }
    }

    public void disableToken(String token) {
        UserDeviceToken deviceToken = userDeviceTokenRepository.findUserDeviceTokenByToken(token)
                .orElseThrow(() -> new NotFoundException("Token=" + token + " not found"));
        deviceToken.setStatus(TokenStatus.INACTIVE);
        userDeviceTokenRepository.save(deviceToken);
    }
}

File: notification-service/src/main/kotlin/ru/javavlsu/kb/notificationservice/service/UserDeviceTokenService.kt
package ru.javavlsu.kb.notificationservice.service

import org.springframework.stereotype.Service
import ru.javavlsu.kb.notificationservice.dto.TokenRegistrationEvent
import ru.javavlsu.kb.notificationservice.exception.UserDeviceTokenNotFoundException
import ru.javavlsu.kb.notificationservice.model.TokenStatus
import ru.javavlsu.kb.notificationservice.model.UserDeviceToken
import ru.javavlsu.kb.notificationservice.repository.UserDeviceTokenRepository

/**
 * UserDeviceTokenService 23.07.2026 thewyolar
 * Copyright (c) 2026.
 */
@Service
class UserDeviceTokenService(
    val userDeviceTokenRepository: UserDeviceTokenRepository
) {

    fun getUserDeviceTokensByUserId(userId: Long): List<UserDeviceToken> {
        return userDeviceTokenRepository.getUserDeviceTokensByUserId(userId)
    }

    fun saveToken(tokenRegistrationEvent: TokenRegistrationEvent) {
        if (!userDeviceTokenRepository.existsByTokenAndUserId(tokenRegistrationEvent.token,
                tokenRegistrationEvent.userId)) {
            val userDeviceToken = UserDeviceToken(userId = tokenRegistrationEvent.userId,
                token = tokenRegistrationEvent.token)
            userDeviceTokenRepository.save(userDeviceToken)
        }
    }

    fun disableToken(token: String) {
        val deviceToken = userDeviceTokenRepository.findUserDeviceTokenByToken(token)
            .orElseThrow { UserDeviceTokenNotFoundException("Token=$token not found") }
        deviceToken.status = TokenStatus.INACTIVE
        userDeviceTokenRepository.save(deviceToken)
    }
}